Allow staff to masquerade as a specific user in the LMS (SOL-816)

smarnach opened the pull request:
**Description**

Enable previewing the LMS as a specific student.  To masquerade as a specific user, select "Specific Student" from the preview dropdown in Courseware or on the Course Info page, then enter the name or email of the user you want to masquerade as and hit enter (or do anything else that blurs the input field).  You can only masquerade as a user enrolled in the current course.  After entering a valid name, the page reloads and is shown with the state of the user you chose.  In contrast to the preview as a generic student, the "Submission History" and "Staff Debug Info" buttons are shown.  The latter can be used to reset the state of the student.

While masquerading, state changes are written to the session when interacting with XBlocks.  This data is reset on page reload, so you can see the student's state again by reloading the page.

There is a [test course on the sandbox](http://sandbox4.opencraft.com/courses/course-v1:OpenCraftX+1+2015/courseware/b5e4cfc740444c2f8a0121412ed137ff/0808bdda7c1a4bf5b3422c4c5ebed90b/).  The standard staff user has instructor priveleges on this course, and the honor user is enrolled as a student.  To test, you can masquerade as the honor user while logged in as staff.  It's easiest to stay logged in as honor in an incognito window at the same time, to make state changes in both windows and see what changes persist.
- - -
**JIRA Story** [SOL-816](https://openedx.atlassian.net/browse/SOL-816)
**Confluence / Product Asset** N/A
**Sandbox URL** [Test course on the sandbox](http://sandbox4.opencraft.com/courses/course-v1:OpenCraftX+1+2015/courseware/b5e4cfc740444c2f8a0121412ed137ff/0808bdda7c1a4bf5b3422c4c5ebed90b/) (Last Updated: 2015-07-08)
**Dependencies** N/A
- - -
**PR Author(s) Notes / To-Do** 
- [ ] Make build pass
- [ ] Add tests

- - -
**Screenshots**
![image](https://cloud.githubusercontent.com/assets/249196/8593243/3cc6b0e0-2636-11e5-8ec4-a635d0316280.png)
- - -
**Reviewers**
Code: @bradenmacdonald, @cpennington; Product: (TBD), UX: @marcotuts

edx-webhook commented:
Thanks for the pull request, @smarnach! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=8775

smarnach commented:
@bradenmacdonald I've set up a sandbox now and updated the testing instructions in the PR description.  I'd appreciate some feedback.

bradenmacdonald commented on lms/djangoapps/courseware/masquerade.py:
```diff
@@ -19,16 +25,21 @@
         self.role = role
         self.user_partition_id = user_partition_id
         self.group_id = group_id
+        self.user_name = user_name
```
What happens if `role` and `user_name` are inconsistent? e.g. `staff` with a student username or `student` with a staff username? (the latter is certainly possible to trigger from the code from what I can see). Shouldn't that be validated?

smarnach replied:
If you use the "Preview as specific student" functionality, the user you enter is automatically treated as a student, regardless of whether the user you entered actually has staff privileges.  I thought about disallowing to masquerade as a different staff user, but I think the current behaviour is good enough.  (Simply setting role to "staff" here when the username entered is a staff user does not work – it would require adjustment in many places in the code.)

bradenmacdonald commented on lms/djangoapps/courseware/masquerade.py:
```diff
@@ -116,3 +174,146 @@ def get_masquerading_group_info(user, course_key):
+        return frag
+    return Fragment(
+        u'Block of type {class_name} not whitelisted for display while masquerading as a'
+        u'specific student is active.'.format(class_name=block.unmixed_class.__name__)
```
This message runs together and says "as aspecific student". I think it could also be softened a bit; perhaps "This type of block cannot be shown while viewing the course as a specific student. ({class_name} does not indicate that it supports read-only views)" ?

smarnach replied:
That wording is certainly better.  As a non-native speaker, I didn't spend much time thinking about a good message, since I learned there are much better qualified people from edX looking over the messages anyway.

bradenmacdonald commented on lms/djangoapps/courseware/masquerade.py:
```diff
@@ -99,6 +126,29 @@ def get_masquerade_role(user, course_key):
+            pass
+        else:
+            masquerade_user.masquerade_settings = user.masquerade_settings
+            del user.masquerade_settings
```
Why do you assign this and then delete it immediately?

smarnach replied:
Masquerading isn't implemented transparently – the masquerading settings are checked explicitly in different places of the code.  One example is the [check in `has_access()` that explicitly calls `is_masquerading_as_student()`](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/access.py#L614).  If I would leave the masquerade settings attached to the `real_user` object, that user still wouldn't be granted staff access.  The `real_user` is used to decide whether to show the "Staff Debug Info" link, and leaving the masquerade settings in place would result in the link not showing up.  This solution seemed the easiest way to achieve the desired behaviour.

I added a comment with a bit of explanation.

bradenmacdonald replied:
Thanks, the comment helps a lot!

bradenmacdonald commented:
This is looking really good. I tested it out and was impressed with the functionality. I have only a couple of comments:

First, I think that the user needs to be better informed about what this mode does. When the user is actively masquerading as a student, I think there should be a yellow banner at the top (under the masquerade settings), that says "You are now viewing the course as [username]. Any changes you make  will not be saved, and will be reset when you return to staff mode."

Second, I experienced a weird bug:
Everything was going fine until I clicked on the Course Discussions tab (while masquerading as 'honor'). I tried submitting a post in order to see who the author would be. The post submission appeared to fail, and then that session somehow got messed up. Now when I try to view the courseware it just says "Page Not Found". The course info page works, but as soon as I click "Courseware", I get a 404.

The info page (works):
![screen shot 2015-07-06 at 4 15 57 pm](https://cloud.githubusercontent.com/assets/945577/8535422/63737ef0-23fa-11e5-84fb-585489f6e0e1.png)

Then I click "courseware":
![screen shot 2015-07-06 at 4 16 14 pm](https://cloud.githubusercontent.com/assets/945577/8535420/5f0f63a6-23fa-11e5-9e0a-abee1428915d.png)

Oddly, the discussion forum is now working again, although my post showed up as "Staff". I don't think the discussion forum is part of your scope, so I'm not too worried about that. It just seems like that might have been what triggered this weird error.

It appears now to be permanently affecting the staff user, so if you go to the sandbox, sign in as staff, view the course, then try to masquerade as "honor@example.com", you may see the issue too.

smarnach commented:
Thanks for the testing.

What you are reporting are actually two completely separate issues.  One is with the discussion tab – it should be completely unaffected by masquerading, but apparently isn't.  The other one is that clicking the Courseware tab results in a 404 (regardless of whether you clicked any other tab before that – you get a 404 even when clicking it while in courseware).  I'll look into the issues and try to fix them.

smarnach commented:
@cpennington Would you be able to review this PR?  It's still not completely finished (there are issues left, and the tests are missing), but I'm going on vacation soon, and would like to get this merged before setting off, the latest by Friday evening, so I'd appreciate feedback as early as possible.

cpennington commented on common/lib/xmodule/xmodule/x_module.py:
```diff
@@ -278,6 +278,10 @@ class XModuleMixin(XModuleFields, XBlock):
 
+    # Whether this module can be displayed in read-only mode.  It is safe to set this to True if
+    # all user state is handled through the FieldData API.
+    show_in_read_only_mode = False
```
Conceptually, this seems like it's more a property of a view than of an XBlock.

smarnach replied:
@cpennington This property is supposed to indicate whether the block stores all its state in XBlock fields, or whether it keeps any state outside of that API, which seems to be a property of a block.  Maybe it just needs a better name?

The original plan was to introduce a new view called `read_only_student_view()`, and only render blocks that implement this view.  The default implementation for blocks that only store state in XBlock fields would be to set `read_only_student_view = student_view`.

This turned out to be more involved in practice.  For one, all blocks with children would need additional machinery to call the right view on the child nodes, but there were additional problems, so I decided to use this much simpler approach.

cpennington replied:
I guess my thought was that an XBlock that stores state externally might still have a view that wouldn't affect that state, and thus should be eligible for masquerading.

But for now, this is fine, I think.

cpennington commented on lms/djangoapps/courseware/masquerade.py:
```diff
@@ -99,6 +126,29 @@ def get_masquerade_role(user, course_key):
+        try:
+            masquerade_user = CourseEnrollment.objects.users_enrolled_in(course_key).get(username=course_masquerade.user_name)
+        except User.DoesNotExist:
+            pass
```
Maybe just `return None` here?

smarnach replied:
Done.

cpennington commented on lms/djangoapps/courseware/masquerade.py:
```diff
@@ -116,3 +174,146 @@ def get_masquerading_group_info(user, course_key):
+    """
+    A wrapper for FieldDataCache used when masquerading as a different user.  Data is read from the
+    underlying cache, but written to the session, so the state of the actual user is not changed.
+    """
```
It would be better to make a `MasqueradingFieldData`, rather than a `MasqueradingFieldDataCache`. Then it would be more applicable to XBlocks, rather than the lms specific machinery.

smarnach replied:
@cpennington I chose to implement the FieldDataCache for two reasons.  First, the `set_score()` method directly modifies state in the database, which we don't want to do when masquerading as a different user.  Second, `field_data_cache` is created at the point where the masquerading is set up and then passed through many layers of function calls before it is actually used, so it was just a lot easier to implement it instead of a `FieldData`, since the creation and initialisation happened right in the place where the masquerading logic was anyway.

Regarding the first point, I just realized that overriding `set_score()` is not enough.  I rather have to make sure that the function `handle_grade_event()`, which calls `set_score()`, isn't called in the first place, since it makes several state-changing calls, so the first reason for my design decision is void.

The second point doesn't hold much water either, since I can simply check whether masquerading is active at the beginning of `get_module_system()`, and instantiate the correct `FieldData` depending on the result, so I don't have to do a lot of refactoring to make this work.

More specifically, I would implement a `MasqueradingKeyValueStore` that takes a generic `KeyValueStore` and a session as arguments, and implements the same logic as the `MasqueradingFieldDataCache`.  This `KeyValueStore` can be turned into a proper `FieldData` using the generic `KvsFieldData`, which is implemented in the XBlock code and is completely independent of the LMS machinery.

Would that be what you had in mind?

smarnach replied:
@cpennington I refactored the code to use a `MasqueradingKeyValueStore` instead.  It required some refactoring in `module_render.py`, but those changes seem safe to me.

I also had to move `set_score()` off the `FieldDataCache`, since the cache is not available anymore at the place where the method is called, and it didn't really use the instance state anyway.  This change feels a bit dubious, since I moved a function from a place where it didn't really fit to a different place where it doesn't really fit.  I'd appreciate suggestions how to improve this.

bradenmacdonald commented on lms/djangoapps/courseware/masquerade.py:
```diff
@@ -99,6 +126,34 @@ def get_masquerade_role(user, course_key):
+            masquerade_user.masquerade_settings = user.masquerade_settings
+            del user.masquerade_settings
+            masquerade_user.real_user = user
+            return masquerade_user
```
Did you consider combining this method with `setup_masquerade`? I see that sometimes `setup_masquerade` is used alone, but this method is only ever used immediately after `setup_masquerade`, and seems to be a natural continuation of the setup process. That would also remove the need to move the `masquerade_settings` from one user object to another, since you could set it on the "right" user immediately. I understand this may not be a good idea, but I just wanted to make sure you've considered it.

smarnach replied:
@bradenmacdonald Yes, I did consider combining those functions, but didn't to keep my changes better confined.  I think there is only one more call to `setup_masquerade()` (for the Course Info page), and it might make sense to replace the user there as well, but since this was outside the scope of this task, I preferred not to change it.

cpennington replied:
Unless the thread ends up particularly long, I'd rather see the other place where masquerading is happening get updated as well, just to keep to a minimum the number of different ways of using this that we're trying to keep working at once.

smarnach replied:
I updated the masquerading code for the Course Info page, and the functionality should be fully working for that page now.  I didn't test it extensively, however, since masquerading only makes a difference in special circumstances on that page.

smarnach replied:
And I also combined the two setup functions now, since it's not necessary to keep them separate anymore.  This enabled handling the edge case that the user we try to masquerade as has been unenrolled in the meantime by simply resetting the masquerading settings.

cpennington commented on lms/djangoapps/courseware/masquerade.py:
```diff
@@ -116,3 +179,83 @@ def get_masquerading_group_info(user, course_key):
+
+        Keys are not JSON-serializable, so we can't use them as keys for the Django session.
+        The implementation is taken from cms/djangoapps/contentstore/views/session_kv_store.py.
+        """
```
Is there enough code in common between this and `session_kv_store.py` to warrant moving it into either `openedx/` or `common/lib/xmodule/`?

smarnach replied:
@cpennington I don't think so.  `session_kv_store.py` both reads from and writes to the session, whereas this key-value store uses the session as a layer on top of some underlying key-value store, which is quite different.  Another difference is that the `session_kv_store.py` implementation dumps the data directly into the session, resulting in a lot of entries with weird keys directly in the session dictionary, which is (a) weird and (b) impossible to clean up.

Even if we fix this weirdness, there would be only a single method (`set()`) with identical implementation, and I don't think that trivial one-liner warrants moving a module around.

cpennington replied:
Ah, that makes sense. I forgot the behavior was different that way. (And yes, not dumping the keys directly into the session is definitely better).

smarnach commented:
@bradenmacdonald I think I fixed the bug with the 404 on the Courseware page.  I tested the discussion forums, but couldn't find any problems there – masquerading doesn't seem to affect them at all, which is the intended behaviour.  I think I also addressed all your other review comments, so please test this again – the sandbox is up to date.

cpennington commented on lms/djangoapps/courseware/model_data.py:
```diff
@@ -955,3 +922,23 @@ def last_modified(self, key):
+
+
+# @contract(user_id=int, usage_key=UsageKey, score="number|None", max_score="number|None")
+def set_score(user_id, usage_key, score, max_score):
```
I like this better, actually. It helps narrow down the `FieldDataCache` interface.

cpennington commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -270,10 +276,15 @@ def get_module_for_descriptor(user, request, descriptor, field_data_cache, cours
         user=user,
         descriptor=descriptor,
-        field_data_cache=field_data_cache,
+        student_data=student_data,
```
Oh, nice. I love that this limits knowledge of the `FieldDataCache` interface even more.

bradenmacdonald commented:
@smarnach I'm getting an error on the sandbox: When masquerading as `honor`, if I try to submit a problem, it's giving me a 500 error. (The AJAX does a POST to `/handler/xmodule_handler/problem_check`, and that view is giving a 500). That was working before in the earlier version.

bradenmacdonald commented:
I still think we should have a yellow notice at the top explaining that changes will not be saved. People may otherwise find it confusing.

smarnach commented:
@bradenmacdonald Sorry about this.  The bug was a stupid typo resulting from copy&paste, and could be fixed by simply removing the method that it was in, since the default implementation is fine for us.  I thought I had tested this though, so not sure why I missed this.  I definitely need to finish the automated tests – manual testing is too error-prone. :-/

I also tested that the session is reset when you change the masquerading settings, so this should be working fine now.  I agree about the yellow banner and will add it.

antoviaque commented:
+1 for adding a warning about the fact that changes are not saved. Note that such a change would need to be reviewed by the UX team, since that will be introducing a new interface element I think - @talbs do you have any recommendations on this, and the bandwidth for reviewing it this week? Sorry for the late notice, we didn't originally think there would be new UX in this story. CC @pbaruah 

@smarnach Btw, the changes made by the student in the meantime (or the instructor when he uses the debug tools to reset the student state for example) aren't shown until you switch back to normal mode, and then back again to "view as specific student". It would be better to refresh on each page load to avoid confusions.

smarnach commented:
@antoviaque I added the yellow banner – it's already live on the sandbox.

You can see changes made by the student immediately, unless you have overridden them by your own changes stored in the session.  I can see that this behaviour is confusing, and I'll reset the session data on page load as you suggested.

talbs commented:
Thanks for the ping, @antoviaque. We're happy to help with reviewing any new patterns with this work. I think @marcotuts has some familiarity with what you folks are doing here. I think he'll be the one jumping in to support things from a UX perspective.

smarnach commented:
@antoviaque I implemented resetting the data on page reloads now, and I think this is indeed a lot less confusing.

andy-armstrong commented:
@catong FYI

smarnach commented:
@braden: I think I've addressed all open comments again, and I tested that the sandbox is working now.  Could you please take another look?

catong commented on lms/templates/courseware/course_navigation.html:
```diff
@@ -34,21 +39,34 @@
                         </select>
-                        <button type="submit" class="sr" name="submit" value="submit">${_("set preview mode")}</button>
+                        <div class="action-preview-username-container">
+                          <label for="action-preview-username" class="action-preview-label">${_("User name or email:")}</label>
```
"User name" -> "Username" (one word)

smarnach replied:
Done.

catong commented on lms/templates/courseware/course_navigation.html:
```diff
@@ -34,21 +39,34 @@
+        % if specific_student_selected:
+        <div class="preview-specific-student-notice">
+          <p>
+            ${_("You are now viewing the course as <i>{user_name}</i>. Any changes you make will not be saved, "
```
How about: "This is a read-only view. Any changes you make are reverted when you refresh the page."

catong replied:
Thinking further about my suggested wording -- I don't want it to sound as if changes made in the Staff Debug Info form are also reverted. Perhaps I should qualify "changes" further: "Changes to... are reverted." @pbaruah @smarnach @antoviaque Thoughts?

smarnach replied:
@catong I was worried that using the term "read-only view" would suggest that people can't really interact with problems on the page, while in fact they can, but thinking about it, I think the next sentence about changes being reverted on refresh makes it pretty clear that you *can* make temporary changes, so I think your first suggestion is good.

To clarify that the debug tools actually do affect student state, I'd suggest to add a clarifying sentence to the debug tools – I guess this will make it more obvious which statement applies to what.

catong replied:
Thanks, @smarnach.
Maybe: "You are now viewing the course as {user_name}. This is a read-only view. Changes you make, other than changes to problem states using the staff debug tools, are reverted when you refresh the page."

smarnach replied:
@catong I pushed a new version to the [sandbox](http://sandbox4.opencraft.com/courses/course-v1:OpenCraftX+1+2015/courseware/b5e4cfc740444c2f8a0121412ed137ff/0808bdda7c1a4bf5b3422c4c5ebed90b/) with the updated strings.  Could you suggest a better message to use inside the debug tools?

bradenmacdonald commented:
@smarnach This is looking quite nice. The yellow bar and resetting the state upon reload are both great improvements. I just noticed a little SASS nit, otherwise this looks great and I'll give you thumbs once the build is passing and you have new tests.

catong commented on lms/djangoapps/courseware/masquerade.py:
```diff
@@ -46,40 +57,71 @@ def handle_ajax(request, course_key_string):
+        except User.DoesNotExist:
+            return JsonResponse({
+                'success': False,
+                'error': 'There is no user with user name or email address {user_name} '
```
"There is no user with ^the^ username or email address <blah> enrolled in this course."

smarnach replied:
Done.

catong commented:
@andy-armstrong Thanks for tagging me for UI string review.
@smarnach I've made some minor edits to the user-visible strings. 
UX should take a look at the yellow banner style: @marcotuts ?
Also, this is existing style, but I keep find the positioning of the "Staff Debug Info" button misleading. The button appears under a line, which seems to separate it from the current problem and brings it very close to the title of the following problem -- confusingly it looks as if it's associated with the problem below instead of the current problem.

marcotuts commented on lms/templates/courseware/course_navigation.html:
```diff
@@ -34,21 +39,34 @@
             </ol>
         </div>
+        % if specific_student_selected:
+        <div class="preview-specific-student-notice">
```
UX: My suggestion here would be to consider this a part of the preview-base element above, which would involve moving this block into <div class="preview-menu">, after the actions <ol>. This way we'd actually not be introducing another visual bar with a different color background.

smarnach replied:
@marcotuts Thanks for the suggestion – I'll try that and we can see how it looks.  The current version is probably a bit too flashy for the overall rather subdued LMS design (although the idea of the yellow bar was to make it stand out).

smarnach replied:
@marcotuts I've pushed a new version to the [sandbox](http://sandbox4.opencraft.com/courses/course-v1:OpenCraftX+1+2015/courseware/b5e4cfc740444c2f8a0121412ed137ff/0808bdda7c1a4bf5b3422c4c5ebed90b/), and I think this actually works quite well.  Could you please take another look?

smarnach commented:
@catong I had exactly the same though regarding the placement of the Staff Debug Info buttons.  While I could change this as part of this PR, it seems out of scope, and should probably be fixed in a separate PR.

marcotuts commented:
Hi everyone!
I drafted some notes from my UX review below, numbering them to hopefully make it easier to reference any replies or comments you may have about them. Thanks! 

Visual Design Notes: 
1.) As mentioned inline in the comments above, it would be great to move the "-notice" div inside of the preview-menu area.
2.) Visually this would convert the support string into a supplemental message within this area instead of a separate bar. 
3.) I would also suggest slightly bumping down the font-size of the notice message, as shown in the screenshot below. 
4.) I've reduced the margin separation between the preview dropdown and the username input from 40px to 20px here as well. 

Screenshot: 
![image](https://cloud.githubusercontent.com/assets/2023680/8580897/3886fbda-258b-11e5-8359-916ecbaa4d9d.png)

Documentation / Labels Discussion: 
There are a few questions / observations I wanted to share and get some feedback on.
5.) Question:  My reaction to the note inside of the staff debug info was that I found it more confusing than helpful unfortunately. Without the label, I would expect staff members to expect all actions in this menu to have an effect on students, but I found myself unclear about the implication of the 'read-only state' described. Are we ok dropping this label entirely and expecting all actions in here to remain consistently active staff debug tools? 
6.) Nit:  I think we should consider having the full explanation of the specific student view linked to in separate documentation, which would let us either drop this note entirely or ensure that the line of text is short and clear (even if it isn't fully explained). The text string makes sense as is, but I just wanted to note that we could easily err on the side of a precise definition, even if really all a user needs to know is that they're now viewing the LMS as another person.

LMS Presence:
7.) Currently this preview bar is only shown on the Courseware tab and Course Info tab, is that correct? I don't think we would expect to have this bar be shown on the Discussion tab or any other part of the LMS for the time being, especially since I don't think the other options in this menu (Preview as Student in Cohort X) work for those other applications yet. 

Thanks all!

marcotuts commented:
One other visual note that may be out of scope  from this PR: 
8.) It would be great to change the Split Text preview control in the LMS to look more like the other Instructor Controls. A quick change would be to wrap it in the same div that the other instructor control elements use, placing the select field inside of this element. Here is a screenshot showing those changes in the browser:
![image](https://cloud.githubusercontent.com/assets/2023680/8581252/94f8eb9c-258d-11e5-9457-ffeb27983f7b.png)

(I'm happy to split this or any of the items above into a separate story as necessary. Thanks!)

catong commented:
The "read only" message looks good brought inside the preview menu area, but I agree with @marcotuts that drawing attention to the ability to make changes, and then having to clarify what changes are saved/not saved in fact adds confusion. Marco, if I understood your suggestion correctly, I agree -- I would be fine with not having any message after selecting a specific student in the "View As" menu, and not having any additional messaging inside the staff debug info dialog.

marcotuts commented:
@catong Yeah, that was my intent with the message above, but I wanted to clarify one small aspect of what I wrote below. 

I do see the value in having a message that confirms for instructors the fact that they are viewing student X's LMS. "You are now previewing the course as if you were audit" or "You are now previewing this course as username".  Without this part of the message you might not know for sure if the control was working.  (Or at least that would be my concern.) If we find that the "Preview as" messaging isn't clear enough in its unobtrusive intent we could revisit that decision and add back in more details to this message perhaps? 

+1 to having the message inside of the staff debug info removed. (Though I understand its original intent and appreciate the effort to clarify the controls!)

catong commented:
@marcotuts thanks for clarifying. @smarnach Sorry for the churn on this! 
How about if we limit the confirmation messaging under the preview menu to the "You are now viewing the course as {username}." piece? I'll comment inline on the files as well. Thanks!

catong commented on lms/templates/courseware/course_navigation.html:
```diff
@@ -34,20 +39,33 @@
+            % if specific_student_selected:
+            <div class="preview-specific-student-notice">
+                <p>
+                    ${_("You are now viewing the course as <i>{user_name}</i>. This is a read-only view. "
```
Limit this message to: "You are now viewing the course as {username}."

catong commented on lms/templates/staff_problem_info.html:
```diff
@@ -79,6 +79,7 @@
       </div>
       <div id="result_${location.name | h}"/>
     </div>
+    <p>${_("Note that these actions actually affect the student state, even if you got to the debug tools from a read-only view.")}</p>
```
Please remove this message.

smarnach commented:
@marcotuts Thanks for the detailed notes.  I've implemented 1. to 4. exactly as you suggested.

5.) I added the note because @catong was worried about this, but I'm completely fine with removing it again.  Thinking about it, I don't actually believe anybody would expect these actions not to have any effect, so I'll simply remove the message again. :-)

6.) I think the preview as specific student is actually a bit more confusing than the other preview mode, so adding a short explanation does make some sense to me, but I'm also fine with shortening it as you suggested.

7.) Yes, that's correct – only these two tabs support preview modes, and we should only add more if we have specific use cases.

8.) This looks quite nice, but I agree that it is outside the scope of this PR. (It would be some work to implement, since it requires some refactoring in the code.  There are some preview modes that don't render the staff-specific elements, so if we simply move the controls there, they would disappear as soon as a user enter such a preview mode.)

marcotuts commented:
Awesome, thanks @catong and @smarnach for the support here, and apologies for the churn as well. All the other notes above for 5-8 make sense as well.

catong commented:
@marcotuts Thanks for the magical UX touch!

smarnach commented:
@catong @marcotuts Thanks for the help!  I think I've now implemented the remaining comments, so please have another look at the sandbox.

catong commented:
Thanks, @smarnach. Looks great to me from the doc/UI strings POV.

catong commented:
@smarnach @pbaruah is it intended that this preview ability only works if a unit is published?

smarnach commented:
@catong I think this is intentional for the existing "View as student" functionality, and it is definitely intentional for the new "View as specific student" functionality.  The idea is that this functioinality helps course staff to see what students would see, to debug issues students report.

There is another preview function, which can be used by clicking the "Preview" button in Studio.  This functionality is completely separate from the masquerading function this PR is about, and it works for unpublished units as well.

catong commented:
OK, thanks for the clarifications @smarnach.

smarnach commented:
Jenkins, test this please.

antoviaque commented:
@catong @smarnach The preview mode isn't setup on the instances, which have the standard fullstack deployment (which doesn't include a preview mode). So on the sandbox both buttons lead to the normal/live mode. 

I haven't checked the code, but from the current behavior, it seem that the "view as student" and "view as specific student" only appear on published units (in the normal/live mode), which seem consistent -- if the user can't see the unit, masquerading as the user to see it doesn't make much sense. 

There could be a case for making it work in the preview mode. Not for units that haven't been published at all yet (if there isn't any data yet for that user, it still doesn't make much sense to see it as he would see it - for example a LibraryContent block wouldn't have selected a random set of exercises for that user yet), but maybe when making changes to an exercise that has already been published, to see how it would impact that user given its current state. It might indeed be the case already, as I think the LMS in preview mode shows the latest draft as the published unit ( @smarnach can you confirm what happens in preview mode for this feature?), but if not imho it seem outside of the original scope to add support for it, as this is a different use case than the debug flow we are trying to support (which is to be able to investigate reports from users about Content Libraries issues only visible to them).

smarnach commented:
@antoviaque @catong I tested the preview feature on my local devstack, and it seems to work fine with unpublished versions of units, i.e. I can see the unpublished version instead of the published version, and use the masquerading settings to view how the unpublished version would look to a generic or specific student.

smarnach commented:
Jenkins, test this please.

catong commented:
@antoviaque @smarnach cc@pbaruah
Ah, I think that was my confusion when trying to test the behavior -- I didn't know that the Preview modes were not working on fullstack instances. Just so I understand, @smarnach you were able to confirm that Preview works both when the course is live and if a course has a release date in the future? (Studio > Settings > Schedule & Details). Thanks!

smarnach commented:
@catong The preview works independent of whether the course is live, yes.  And in preview mode, you can even masquerade as a student before the course is live.  Only when viewing the live version, you can't masquerade as a student if the course is not live.

smarnach commented:
@cpennington Would you be able to have a look at this PR again?  There were a few further comments, and adding tests took me a bit longer than expected.  It would be great if we could get this merged today anyway, since I'm going on vacation tomorrow.

I can completely understand if this is not possible on such a short time scale, but if it is, it would save me the effort of handing this over to someone else at OpenCraft. :)

smarnach commented:
@bradenmacdonald Feel free to take a final look as well.

catong commented on lms/djangoapps/courseware/masquerade.py:
```diff
@@ -116,3 +164,79 @@ def get_masquerading_group_info(user, course_key):
+        return frag
+    return Fragment(
+        u'This type of block cannot be shown while viewing the course as a specific student. '
+        u'({class_name} does not indicate that it supports read-only views.)'.format(
```
@antoviaque @smarnach @pbaruah @marcotuts Do you think we can omit this 2nd sentence? I don't think it's the kind of information that course team members need to know.

smarnach replied:
Omitting the second sentence is perfectly reasonable.

catong commented on lms/djangoapps/courseware/masquerade.py:
```diff
@@ -116,3 +164,79 @@ def get_masquerading_group_info(user, course_key):
+    if getattr(block, 'show_in_read_only_mode', False):
+        return frag
+    return Fragment(
+        u'This type of block cannot be shown while viewing the course as a specific student. '
```
Suggest a different term than "block" in this message, since the audience is course team members. Would "problems" be accurate?

smarnach replied:
@catong "Problem" might be too limited in scope.  I think the term used in the edX documentation term is "component": http://edx.readthedocs.org/en/latest/glossary.html#c

catong replied:
Thanks, @smarnach. Let's go with "This type of component cannot be shown while viewing the course as a specific student." 
@pbaruah @marcotuts ?

smarnach replied:
Done.

cpennington commented on lms/djangoapps/courseware/masquerade.py:
```diff
@@ -106,6 +146,14 @@ def is_masquerading_as_student(user, course_key):
+    Returns whether the user is a staff member masquerading as a specific student.
+    """
+    course_masquerade = get_course_masquerade(user, course_key)
+    return bool(course_masquerade and course_masquerade.user_name)
```
Nit: maybe `CourseMasquerade.__nonzero__` should just check `course_masquerade.user_name`

smarnach replied:
`CourseMasquerade` can describe different kinds of masquerading configuration: as generic student, as part of some content group or as a specific student.  Singling out masquerading as a specific student as `True` and considering everything else as `False` seems unnatural.

There are other places in the code that check `course_masquerade and course_masquerade.group_id`.  The conjunction is necessary since `course_masquerade` might be `None`.

cpennington commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -673,8 +673,8 @@ class TestAccordionDueDate(BaseDueDateTests):
-            self.request, course, course.get_children()[0].scope_ids.usage_id.to_deprecated_string(),
-            None, None
+            self.request.user, self.request, course,
+            course.get_children()[0].scope_ids.usage_id.to_deprecated_string(), None, None
```
Nit: `unicode(...)` rather than `...to_deprecated_string()`.

smarnach replied:
Done.

cpennington commented:
:+1:

bradenmacdonald commented:
:+1:

catong commented:
:+1: Thanks!

marcotuts commented:
Thanks for updating the sandbox! :+1: from a UX standpoint.

bradenmacdonald commented:
jenkins run python please. Timeout caused build failure.

bradenmacdonald commented:
Squashed and rebased. Old commits can be seen at c5426b8d016b02d98157e1f713ce39288c7a3a85. 

@smarnach is away so I am merging this now for him. Thanks to all the reviewers.

e-kolpakov commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -430,7 +439,7 @@ def handle_grade_event(block, event_type, event):  # pylint: disable=unused-argu
         max_grade = event.get('max_value')
 
-        field_data_cache.set_score(
+        set_score(
```
@cpennington @smarnach I know this is a bit outdated, but there's possible bug in this method now. 

So, since `set_score` does not assert on `user_id` being equal to current user id, it would be possible to send grade events with other users' ids. This might or might not be intentional, and in both cases there are problems:

1. If it is intentional, than  there's a problem with `_fulfill_content_milestones` invocation a couple of lines below, as it takes `user` from outer scope, rather than getting it by id. I'm not sure if that could cause trouble already, but still it *could* cause issues in future.
2. If it is not, than some validation/assertion should be added to avoid that, or, even better, `user_id` parameter should be removed from `grade` event.

Let me know what you think.

If you're interested in why I'm getting to this code: this commit is not there yet in solutions, and I've tried sending grade for other user. Resulted in [exception|https://github.com/open-craft/xblock-group-project-v2/pull/64#issuecomment-158072754]. I ended up removing the code that caused it, but it would be still good to know, if sending events for other users in is ok in general and for `grade` event in particular.

cpennington replied:
No, we should only be sending grades for the current user. However, I think that assertion will have to live outside the `set_score` function, since that function doesn't have any way of knowing who the current user is.

e-kolpakov replied:
@cpennington maybe we should remove the `user_id` parameter from event than? This would help avoiding confusion in the future and  simplify the code, instead of adding a new check (and hence, adding more code).
